### DISCUSSION TITLE
03_pytorch_computer_vision.ipynb crossEntropyLoss()

### DISCUSSION THREAD

**Discussion by rithvikshetty** [mrdbourke/pytorch-deep-learning#1031] [Q&A]
![Screenshot from 2024-08-05 00-10-50](https://github.com/user-attachments/assets/b234904d-8d1b-4fd3-8d27-c877d3098bd4)

Model used in the video for this was:

```
class FashionMNISTModelV1(nn.Module):
    def __init__(self, input_shape: int, hidden_units: int, output_shape: int):
        super().__init__()
        self.layer_stack = nn.Sequential(
            nn.Flatten(), # flatten inputs into single vector
            nn.Linear(in_features=input_shape, out_features=hidden_units),
            nn.ReLU(),
            nn.Linear(in_features=hidden_units, out_features=output_shape),
            nn.ReLU()
        )
    
    def forward(self, x: torch.Tensor):
        return self.layer_stack(x)
```

But, the [CrossEntropy](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) requires the input to be Logits and not the activated output right?

So, shouldn't we remove the ReLU from the last output layer?

**Comment by LuluW8071** [comment]
Yes, the activation function should be kept only between layers not outside the layers that is the case for every model architecture

```python
class FashionMNISTModelV1(nn.Module):
    def __init__(self, input_shape: int, hidden_units: int, output_shape: int):
        super().__init__()
        self.layer_stack = nn.Sequential(
            nn.Flatten(), # flatten inputs into single vector
            nn.Linear(in_features=input_shape, out_features=hidden_units),
            nn.ReLU(),
            nn.Linear(in_features=hidden_units, out_features=output_shape),
        )
    
    def forward(self, x: torch.Tensor):
        return self.layer_stack(x)
```

**Reply by rithvikshetty** [reply]
In that case the input for `BCEWithLogitsLoss` won't be the logits right? It'd be the non-linear output.

**Reply by LuluW8071** [reply]
Inputs for `BCEWithLogitsLoss` are logits. It wont be called non-linear either rather called __raw and unbounded logits__  to be precise. In case of `BCEWithLogitsLoss`, when you feed these __raw logits__, the loss internally applies sigmoid function converting it into probabilities and calculates loss_fn between preds and ground truth.

**Reply by rithvikshetty** [reply]
Exactly, when you apply bounded activation function (sigmoid in this case) for the model used in the video
```
class FashionMNISTModelV1(nn.Module):
    def __init__(self, input_shape: int, hidden_units: int, output_shape: int):
        super().__init__()
        self.layer_stack = nn.Sequential(
            nn.Flatten(), # flatten inputs into single vector
            nn.Linear(in_features=input_shape, out_features=hidden_units),
            nn.ReLU(),
            nn.Linear(in_features=hidden_units, out_features=output_shape),
            nn.ReLU()
        )
    
    def forward(self, x: torch.Tensor):
        return self.layer_stack(x)
```

the output is activated, to which you are again applying sigmoid in `BCEWithLogitsLoss`.

**Reply by rithvikshetty** [reply]
> It'd be the non-linear output

I meant that, if we use the model from the video, we'd not be using logits, but the bounded output.

**Reply by LuluW8071** [reply]
**Do Note:**

- It is not recommended to use activation functions such as ReLU, SiLU, GeLU, Leaky ReLU, or others in the final layers of a model architecture (except for Sigmoid and Softmax, depending on the loss function u choose). In ur case, applying ReLU in the final layer is not appropriate as it will restrict the output to the range from 0 to positive values, which is not suitable for all tasks.

> If the video has ReLU() on final layers, it is still not recommended.

**Reply by LuluW8071** [reply]
> I meant that, if we use the model from the video, we'd not be using logits, but the bounded output.

The final layer outputs are always logits for BCEwithLogitLoss and CrossEntropyLoss except for BCELoss (in BCELoss u should apply sigmoid in final layer)

Check documentation of pytorch once for further clarification.
